README.md
# New Features Guideline (新功能指引)

[![Version npm][version]](https://www.npmjs.com/package/guideline?activeTab=versions)
[![Build Status][build]](https://travis-ci.org/johvin/guideline)
[![Dependencies][david]](https://david-dm.org/johvin/guideline)
[![Known Vulnerabilities][vulnerabilities]](https://snyk.io/test/npm/guideline)
[![License][license]](https://opensource.org/licenses/MIT)

[version]: https://img.shields.io/npm/v/guideline.svg?style=flat-square
[build]: http://img.shields.io/travis/johvin/guideline/master.svg?style=flat-square
[david]: https://img.shields.io/david/johvin/guideline.svg?style=flat-square
[vulnerabilities]: https://snyk.io/test/npm/guideline/badge.svg?style=flat-square
[license]: https://img.shields.io/badge/License-MIT-brightgreen.svg?style=flat-square

> This is a simple guideline tool based on which you can quickly implement new features guidelines for website users.

## Usage

There are two ways to play a guideline, a simple way and an advanced way. The simple way is an encapsulation of the advanced way which can do more setting and control the playing of guideline.

*Guideline also supports using keyboard shortcuts to control playback. See the relevant part of API for details*

### Simple way

```js
const guideline = require('guideline');

const visitTimes = parseInt(localStorage.getItem('visitTimes')) || 0;

if (visitTimes === 0) {
  const guideOptions = [{
    content: 'Welcome, the new features guidelines come online !'
  }, {
    element: document.getElementById('notificationwrapper'),
    content: 'all system messages and notification is here',
    style: 'font-size: 20px; color: red;',
    position: 'top'
  }, {
    element: document.querySelector('.datepicker-hint'),
    content: 'datepicker hint will tell you the datepicker\'s date range restriction',
    position: 'bottom'
  }];

  guideline(guideOptions, function () {
    console.log('guideline is over');
  });
}
```

### Advanced way

```js
const guideline = require('guideline');

// something else ...

const guideOptions = [{
  content: 'Welcome, the new features guidelines come online !'
}, {
  element: document.getElementById('notificationwrapper'),
  content: 'all system messages and notification is here',
  style: 'font-size: 20px; color: red;',
  position: 'top'
}, {
  element: document.querySelector('.datepicker-hint'),
  content: 'datepicker hint will tell you the datepicker\'s date range restriction',
  position: 'bottom'
}];

const gl = new guideline.Guideline(guideOptions, function (total, played) {
  console.log('guideline is over, total valid ', total, 'played ', played);
});

// set hint text maximum width
gl.hintTextMaxWidth = 800
// set hint text font size
gl.hintFontSize = 20

// start the guideline
gl.play();

// autoplay the next hint every 3 seconds
const timer = setInterval(function () {
  if (gl.hasNext()) {
    gl.next();
  } else {
    clearInterval(timer);
    gl.stop();
  }
}, 3000);
```

## API

### guideline(configuration, callback)

This is the simple way to play a guideline which accepts two parameters whose description are listed in the relevant sections as follows.

### guideline.Guideline Constructor (configuration, callback)

This is the advanced way to play a guideline which parameters are the same as `guideline(configuration, callback)`. It will return a Guideline instance based on whose prototype's method you can do more control during the playing.

### configuration - guideline(configuration)

The configuration item should be an array, each of which is an object.

#### Configuration Subitem Object

Properties:
- element (HTMLElement, optional) : the guided element. when `element` is null, the guideline text is centered relative to the browser window.
- content (string, required) : the guideline text.
- position (string, optional, defaults to 'bottom') : used to set the location of the guideline text relative to the guided element. enumerated type，value is `'top'` or `'bottom'`.
- style (string, optional) : custom style for the guideline text.

### callback(total, playedAmount) - guideline(configuration, callback)

`callback` is a function, which will be invoked when the guideline stops. `callback` takes two parameters `total` and `playedAmount`, representing the total amount of valid hints and the actual playback amount of hints.

### Guideline instance properties

- hintTextMaxWidth (number, defaults to 400) : used to limit the hint text maximum width, unit is pixel.
- hintFontSize (number, defaults to 24) : used to define the hint text font size, unit is pixel.

### Guideline instance method

- play : start the guideline after doing some settings
- prev : play the previous hint
- next : play the next hint, if there's no more hint, stop the guideline and exit
- hasNext : tell if there is a follow-up hint
- stop : stop playing the guideline

### Keyboard shortcuts

Guideline supports some keyboard shortcuts to play the next or previous hint as follows.

| keyboard | action |
| --- | --- |
| arrow right | play next |
| enter | play next |
| space | play next |
| arrow left | play previous |
| backspace | play previous |
| esc | stop playing |

## CHANGELOG

[`CHANGELOG`](./CHANGELOG.md)

## License

MIT

## References

- http://ejulianova.github.io/guides/


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2018-04-26
### Added
- hasNext api can tell if there is a follow-up hint

### Fixed
- fix spelling mistakes in readme

### Changed
- callback function takes two parameters `total` and `playedAmount`, representing the total amount of valid hints and the actual played amount of hints


## [0.2.1] - 2018-04-20
### Fixed
- fix issue 1 that mask size does not cover the whole page when the page's height is bigger than window.innerHeight
- forbid scrolling page when guideline is playing
- fix the bug that the top and bottom edge of the arrow is overflow when it is skew
- fix the bug that the arrow direction reverses left and right if the guideline text position is on the top of the guided element
- do not play a guideline if the configuration is empty array

### Changed
- update changelog
- update package keywords and description
- set hint text line-height to 1.25


## [0.2.0] - 2018-03-23
### Added
- add an arrow shape from the guide text to the guided element
- add release script in `build` directory
- add Change log
- support umd modular solution

### Changed
- update the setting of webpack configuration
- update package.json to add "bugs" item and add "release" script command
- remove unnecessary published files, such as yarn.lock, webpack.config.js and etc.
- update README.md to add API instructions


## [0.1.0] - 2018-03-20
### Added
- the configuration supports not setting the guided element, in which case the guideline text is centered relative to the browser window
- highlight the guided element with the animation of breathing
- support setting custom style of the guideline text
- support setting the location of the guideline text relative to the guided element, top or bottom
- calculate the location of the guideline text automatically


webpack.config.js
var path = require('path');
var webpack = require('webpack');
var SmartBannerPlugin = require('smart-banner-webpack-plugin');
var pkg = require('./package.json');

module.exports = {

  entry: {
    index: './index.js',
  },

  output: {
    path: path.resolve(__dirname, './dist'),
    filename: "guideline.min.js",
    pathinfo: false,
    libraryTarget: 'commonjs2',
    library: 'guideline'
  },

  target: 'web',

  module: {
    rules: [
      {
        test: /\.js/,
        loader: 'babel-loader',
        query: {
          babelrc: false,
          presets: [
            // A Babel preset that can automatically determine the Babel plugins and polyfills
            // https://github.com/babel/babel-preset-env
            ['env', {
              targets: {
                browsers: [
                  ">1%",
                  "last 4 versions",
                  "Firefox ESR",
                  "not ie < 9"
                ],
              },
              modules: false,
              useBuiltIns: false,
              debug: false,
            }],
            // Experimental ECMAScript proposals
            // https://babeljs.io/docs/plugins/#presets-stage-x-experimental-presets-
            'stage-2'
          ],
        }
      }
    ]
  },

  plugins: [
    new webpack.optimize.UglifyJsPlugin({
      sourceMap: false,
      compress: {
        warnings: true
      },
      output: {
        // remove file comment (webpack.BannerPlugin)
        comments: false,
        screw_ie8: true,
      }
    }),
    new SmartBannerPlugin({
      banner: `[filename] v${pkg.version}\n\nAuthor: johvin\nDate: ${new Date().toISOString()}`,
      raw: false,
      entryOnly: true
    })
  ],

  stats: {
    colors: true,
    reasons: false,
    hash: false,
    version: false,
    timings: true,
    chunks: false,
    chunkModules: false,
    cached: false,
    cachedAssets: false
  }

};

index.js
// judge if the browser event support passive mode
let passiveSupported = false;

try {
  const options = Object.defineProperty({}, 'passive', {
    get: function () {
      passiveSupported = true;
    }
  });
  window.addEventListener('test', null, options);
} catch (e) {}

function isArray(arr) {
  return Object.prototype.toString.call(arr) === '[object Array]';
}

function Guideline(guideOptions, callback) {
  // hint text font-size setting
  this.hintFontSize = 24;
  this.hintTextMaxWidth = 400;
  this.guideOptions = isArray(guideOptions)
    ? guideOptions.filter(it => typeof it === 'object' && typeof it.content === 'string' && it.content.trim().length > 0)
    : [];
  this.callback = callback;
  this.guideElement = null;
  this.cssInjected = false;
  this.htmlInjected = false;
  this.isGuiding = false;
  this.guideIndex = -1;
  this.resizeWindowHandler = this.adjustGuideHintRootElementPosition.bind(this);
  this.clickHandler = this.clickHandler.bind(this);
  this.keyUpHandler = this.keyboardHandler.bind(this);
}

Guideline.prototype = {
  constructor: Guideline,

  // start play the guideline
  play: function play() {
    const { length } = this.guideOptions;

    if (length > 0 && !this.isGuiding) {
      this.isGuiding = true;
      this.guideIndex = -1;

      this.injectCss();
      this.injectHTML();
      this.next();
    }
  },

  // play the previous guide
  prev: function prev() {
    if (!this.isGuiding) return;
    const currentIndex = this.guideIndex - 1;

    if (currentIndex >= 0) {
      this.show(currentIndex);
    }
  },

  // play the next guide
  next: function next() {
    if (this.hasNext()) {
      this.show(this.guideIndex + 1);
    } else if (this.isGuiding) {
      this.stop();
    }
  },

  // tell if there is a follow-up
  hasNext: function hasNext() {
    if (!this.isGuiding) return false;

    const total = this.guideOptions.length;
    const nextIndex = this.guideIndex + 1;

    return nextIndex >= 0 && nextIndex < total;
  },

  // show specific guide
  show: function (guideIndex) {
    const curHintOpt = this.currentHintOption = this.guideOptions[guideIndex];

    if (this.guideElement) {
      this.guideElement.classList.remove('guideline-current-element');
    }
    this.guideElement = curHintOpt.element;
    if (this.guideElement) {
      this.guideElement.classList.add('guideline-current-element');
    }

    // guide hint root element's relative position of current guide element
    this.guideRelativePosition = curHintOpt.position;
    if (this.guideRelativePosition !== 'bottom' && this.guideRelativePosition !== 'top') {
      this.guideRelativePosition = 'bottom';
    }

    if (!this.guideHintRootElement) {
      this.guideHintRootElement = this.guideRootElement.querySelector('.guideline-guide');
    }

    this.adjustGuideHintRootElementPosition();

    // update hint content and style
    const hintElem = this.guideHintRootElement.firstElementChild;
    if (hintElem) {
      hintElem.textContent = curHintOpt.content;
      hintElem.style.cssText = 'style' in curHintOpt ? curHintOpt.style : '';
    }

    this.guideIndex = guideIndex;
  },

  // adjust guide content root element's position
  // according to current guide element's position
  adjustGuideHintRootElementPosition: function () {
    if (this.guideElement) {

      // hint is at the bottom of the guide element
      const isBottom = this.guideRelativePosition === 'bottom';

      // Making guideElement scroll into view before invoking getBoundingClientRect
      // can cover the case that the guideElement's parent or grandfather has a fixed position which leads to
      // guideline's height equals to window.innerHeight, in the meanwhile the guideElement's computed top is greater than window.innerHeight
      // in which case the guide hint element will be outside of the screen.
      if (this.guideElement.scrollIntoViewIfNeeded) {
        this.guideElement.scrollIntoViewIfNeeded();
      } else {
        // note: some browser do not support scrollIntoViewIfNeeded() now.
        this.guideElement.scrollIntoView(isBottom);
      }
      const {
        left,
        right,
        top,
        bottom
      } = this.guideElement.getBoundingClientRect();

      const winWidth = window.innerWidth;
      // hint text padding
      const textPadding = 10;
      let hintWidth = this.calcHintTextWidth(this.hintFontSize, this.currentHintOption.content) + textPadding * 2;

      hintWidth = Math.ceil(hintWidth);
      if (hintWidth > this.hintTextMaxWidth) {
        hintWidth = this.hintTextMaxWidth;
      }

      // guide element x center
      const elemXCenter = (left + right) >> 1;
      // calc hint text x center 
      let hintXCenter = elemXCenter;
      if (hintXCenter + hintWidth / 2 > winWidth) {
        hintXCenter = Math.floor(winWidth - hintWidth / 2);
      }

      if (hintXCenter - hintWidth / 2 < 0) {
        hintXCenter = Math.ceil(hintWidth / 2);
      }

      // hint root element left coordinatee
      const hintLeft = Math.floor(hintXCenter - hintWidth / 2);
      // svg line coordinate
      const linePos = {};
      const svgInnerHeight = 50;
      const svgPaddingVertical = 10; // padding: 10px 0;
      const svgOuterHeight = svgInnerHeight + svgPaddingVertical * 2;

      linePos.x1 = hintWidth >> 1;
      linePos.x2 = elemXCenter - hintLeft;
      if (isBottom) {
        linePos.y1 = svgInnerHeight;
        linePos.y2 = 0;
      } else {
        linePos.y1 = 0;
        linePos.y2 = svgInnerHeight;
      }

      const verticalPosition = isBottom ? `top: ${bottom + 10}px` : `bottom: ${window.innerHeight - top + 10}px`;
      const paddingVertical = `padding-${isBottom ? 'top' : 'bottom'}: ${svgOuterHeight}px`;
      this.guideHintRootElement.style.cssText = `${paddingVertical}; left: ${hintLeft}px; width: ${hintWidth}px; ${verticalPosition};`;

      // svg setting
      const svg = this.guideHintRootElement.querySelector('svg');
      const line = svg.querySelector('line');
      svg.setAttribute('width', hintWidth);
      svg.setAttribute('height', svgOuterHeight);

      if (isBottom) {
        svg.style.top = 0;
        svg.style.bottom = 'auto';
      } else {
        svg.style.top = 'auto';
        svg.style.bottom = 0;
      }

      Object.keys(linePos).forEach((attr) => {
        line.setAttribute(attr, linePos[attr]);
      });
    } else {
      this.guideHintRootElement.style.cssText = `left: 50%; top: 50%; transform: translate(-50%, -50%);`;
      const svg = this.guideHintRootElement.querySelector('svg');
      svg.setAttribute('width', 0);
      svg.setAttribute('height', 0);
    }
  },

  // calculate the layout width of hint text
  calcHintTextWidth: function (fontSize, text) {
    const span = document.createElement('span');
    span.style.position = 'absolute';
    span.style.left = '-10000px';
    span.style.bottom = 0;
    span.style.display = 'inline-block';
    span.style.visibility = 'hidden';
    span.style.fontSize = fontSize + 'px';
    span.innerText = text;

    document.body.appendChild(span);
    const rect = span.getBoundingClientRect();
    document.body.removeChild(span);

    return rect.width;
  },

  // inject css into document's head
  injectCss: function injectCss() {
    if (this.cssInjected) {
      return;
    }

    const cssId = 'guideline-css-style';
    let elem = document.getElementById(cssId);

    if (!elem) {
      elem = document.createElement('style');
      elem.setAttribute('type', 'text/css');
      elem.id = cssId;

      const cssText = `
        .guideline-current-element {
          position: relative;
          z-index: 9991;
          box-shadow: 0 0 4px 4px rgba(255, 0, 0, 0.8);
          animation: breathe 2s ease 1s infinite;
        }

        .body-with-guideline {
          overflow: hidden;
          height: 100%;
        }

        .guideline-wrapper,
        .guideline-bg,
        .guideline-mask {
          position: fixed;
          top: 0;
          left: 0;
          right: 0;
          bottom: 0;
        }

        .guideline-wrapper {
          transition: opacity 0.5s ease;
          opacity: 0;
        }

        .guideline-wrapper.show {
          opacity: 1;
        }

        .guideline-bg {
          background: hsla(0, 0%, 0%, 0.6);
          z-index: 9990;
        }

        .guideline-mask {
          z-index: 9992;
        }

        .guideline-guide {
          position: absolute;
          color: hsl(0, 0%, 100%);
          font-size: ${this.hintFontSize}px;
          z-index: 9993;
        }

        .guideline-guide > .guideline-hint {
          display: inline-block;
          padding: 10px;
          line-height: 1.25;
          text-shadow: 0 0 20px currentColor;
        }

        @keyframes breathe {
          from {
            box-shadow: 0 0 4px 4px rgba(255, 0, 0, 0.8);
          }
          50% {
            box-shadow: 0 0 2px 1px rgba(255, 0, 0, 0.6);
          }
          to {
            box-shadow: 0 0 4px 4px rgba(255, 0, 0, 0.8);
          }
        }
      `;

      // ie8 及之前版本不支持，可使用 cssText 兼容
      elem.textContent = cssText;
      document.head.appendChild(elem);
    }

    this.cssInjected = true;
  },

  // remove css from document's head
  unloadCss: function unloadCss() {
    if (this.cssInjected) {
      const css = document.getElementById('guideline-css-style');
      css.parentNode.removeChild(css);
    }
  },

  // inject guide html fragment into document's body
  injectHTML: function injectHTML() {
    if (this.htmlInjected) {
      return;
    }

    // if there already exists a guideline-wrapper, throw error
    if (document.querySelector('.guideline-wrapper')) {
      throw new Error('A guideline already exists, it\'s not allowed to run another guideline until the previous one finishes');
    }

    this.htmlInjected = true;

    const html = `
      <div class="guideline-wrapper show">
        <div class="guideline-bg"></div>
        <div class="guideline-mask"></div>
        <div class="guideline-guide">
          <span class="guideline-hint"></span>
          <svg width="0" height="0" style="padding: 10px 0; position: absolute; left: 0; box-sizing: border-box;">
            <defs>
              <marker id="arrow" viewBox="0 0 12 12" refX="10" refY="6" markerWidth="6" markerHeight="20" orient="auto" style="fill: currentColor;">
                <path d="M 0 0 L 12 6 L 0 12 z" />
              </marker>
            </defs>
            <line x1="0" y1="0" x2="100" y2="100" style="stroke:currentColor;stroke-width:2" marker-end="url(#arrow)"/>
          </svg>
        </div>
      </div>
    `;

    document.body.insertAdjacentHTML('beforeend', html);
    // 阻止 body 滚动
    document.body.classList.add('body-with-guideline');

    this.guideRootElement = document.querySelector('.guideline-wrapper');
    // bind click listener
    this.guideRootElement.addEventListener('click', this.clickHandler, passiveSupported ? { passive: true } : false);
    // bind keyboard event listener
    window.addEventListener('keyup', this.keyUpHandler, true);
    // listen window resize event
    window.addEventListener('resize', this.resizeWindowHandler, passiveSupported ? { passive: true } : false);
  },

  // remove guide html fragment from document's body
  unloadHTML: function unloadHTML() {
    if (!this.htmlInjected) {
      return;
    }
    this.htmlInjected = false;

    document.body.classList.remove('body-with-guideline');

    if (this.guideRootElement) {
      this.guideRootElement.parentNode.removeChild(this.guideRootElement);
      this.guideRootElement = null;
    }

    // removeEventListener or detachEvent
    // http://www.runoob.com/jsref/met-element-removeeventlistener.html
    // useCapture 是否必须分情况而定
    //
    // bind keyboard event listener
    window.removeEventListener('keyup', this.keyUpHandler);
    // unbind window resize listener
    window.removeEventListener('resize', this.resizeWindowHandler);
  },

  // handle click event to navigate guideline
  clickHandler: function clickHandler(e) {
    e.stopPropagation();
    this.next();
  },

  // handle keyboard event
  keyboardHandler: function keyboardHandler(e) {
    if (e.ctrlKey || e.altKey || e.shiftKey || e.metaKey) {
      // 组合键，不是要控制功能指引的前进后退
      return;
    }

    switch (e.keyCode || e.which) {
      // arrow right
      case 39:
      // enter
      case 13:
      // space
      case 32:
        this.next();
        break;
      // arrow left
      case 37:
      // backspace
      case 8:
        this.prev();
        break;
      // esc
      case 27:
        this.stop();
        break;
      default:
    }
  },

  // stop play the guideline
  stop: function stop() {
    if (this.isGuiding) {
      this.isGuiding = false;

      const total = this.guideOptions.length;
      const playLength = this.guideIndex + 1;
      this.guideIndex = -1;

      this.destroy();

      if (this.callback && typeof this.callback === 'function') {
        this.callback(total, playLength);
      }
    }
  },

  destroy: function destroy() {
    this.guideOptions = null;
    if (this.guideElement) {
      this.guideElement.classList.remove('guideline-current-element');
      this.guideElement = null;
    }

    this.unloadHTML();
    this.unloadCss();
  }
};

function guide(guideOptions, callback) {
  const gl = new Guideline(guideOptions, callback);
  gl.play();
};

guide.Guideline = Guideline;

module.exports = guide;
